File: java/cours/src/session2/td/heritageMultiple/AlgebreNatRecursifZero.java
package session2.td.heritageMultiple;

import session1.td.Nat;

public interface AlgebreNatRecursifZero extends Nat{

	@Override
	public default Nat zero() {
		return this.creerZero();
	}

	@Override
	public default Nat somme(Nat x) {
		return x;
	}

	@Override
	public default Nat un() {
		return this.creerSuccesseur(this.creerZero());
	}

	@Override
	public default Nat produit(Nat x) {
		return this.creerZero();
	}

	@Override
	public default Nat modulo(Nat x) {
		return this.creerZero();
	}

	@Override
	public default Nat div(Nat x) {
		return this.creerZero();
	}

	public default String representer() {
		return "0";
	}

	public default boolean estEgal(Object obj) {
		if (!(obj instanceof Nat))
			return false;
		Nat x = (Nat) obj;
		return x.estNul();
	}
}

File: typescript/session2/heritageMultiple/AlgebreNatRecursifZero.ts
import {Nat} from "../../session1/td/Nat/Nat";

export abstract class AlgebreNatRecursifZero extends Nat{

    zero() : Nat{
        return this.creerNatZero();
    }

    somme(x :Nat) : Nat {
        return x;
    }

    un() : Nat{
        return this.creerNatSucc(this.creerNatZero());
    }

    produit(x :Nat) : Nat{
        return this.creerNatZero();
    }

    modulo(x :Nat) : Nat{
        return this.creerNatZero();
    }

    div(x : Nat) : Nat{
        return this.creerNatZero();
    }

    representer() : string{
        return "0";
    }

    estEgal(obj : Object) : boolean {
        if (!(obj instanceof Nat))
            return false;
        let x = <Nat> obj;
        return x.estNul();
    }
}
